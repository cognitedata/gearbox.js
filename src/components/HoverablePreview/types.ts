import React from 'react';

export type TBorders = ('top' | 'right' | 'bottom' | 'left')[];
export type TAlign = 'left' | 'right' | 'center';

interface HPBasic {
  className?: string;
}

export interface HPIconProps extends HPBasic {
  onMouseOver?: React.MouseEventHandler;
  onMouseEnter?: React.MouseEventHandler;
  onMouseLeave?: React.MouseEventHandler;
  onClick?: React.MouseEventHandler;
  style?: React.CSSProperties;
}
export interface HPProps extends HPBasic {
  displayOn?: 'hover' | 'click' // when ommited, hoverable preview is always visible and HoverIcon is hidden
  noShadow?: boolean;
  fadeIn?: boolean;
  show?: boolean;
  children: React.ReactNode | string;
}
export interface HPCellProps extends HPBasic {
  children: React.ReactNode | string;
  title?: string;
  align?: TAlign;
  borders?: TBorders;
}

