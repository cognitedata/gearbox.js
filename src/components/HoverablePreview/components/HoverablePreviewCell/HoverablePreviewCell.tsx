import React from 'react';
import styled, { css } from 'styled-components';
import { HPCellProps, TBorders } from '../../types';

const adjustBorders = (borders?: TBorders) => {
  const top = borders && borders.includes('top') ? 'solid' : 'none';
  const right = borders && borders.includes('right') ? 'solid' : 'none';
  const bottom = borders && borders.includes('bottom') ? 'solid' : 'none';
  const left = borders && borders.includes('left') ? 'solid' : 'none';

  return css`
    border: 1px solid #d9d9d9;
    border-style: ${top} ${right} ${bottom} ${left};
  `;
};

const StyledHPCell = styled.div<HPCellProps>`
  display: flex;
  flex-direction: column;
  padding: 16px;
  box-sizing: border-box;
  color: #262626;
  font-size: 14px;
  line-height: 24px;
  width: ${({ align }) => (align ? '50%' : '100%')};
  ${({ align }) =>
    align === 'right' &&
    css`
      margin-left: auto;
    `}
  ${({ borders }) => borders && adjustBorders(borders)}

    .title {
    text-transform: uppercase;
    color: #595959;
    font-size: 12px;
    line-height: 20px;
  }
`;

export class HoverablePreviewCell extends React.Component<HPCellProps, {loaded: boolean}> {
  render() {
    const { title, children, borders, align } = this.props;

    return (
        <StyledHPCell className="hp-cell" borders={borders} align={align}>
          {title && <div className="hp-cell-title">{title}</div>}
          {children}
        </StyledHPCell>
    );
  }
}
