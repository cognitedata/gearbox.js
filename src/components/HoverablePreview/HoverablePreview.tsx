import React from 'react';
import styled, { css } from 'styled-components';
import { HoverablePreviewCell } from './components/HoverablePreviewCell';
import { HoverablePreviewHeader } from './components/HoverablePreviewHeader';
import { HoverablePreviewProps } from './types';

export const StyledHoverablePreview = styled.div<HoverablePreviewProps>`
  width: 480px;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  color: #111;
  word-break: break-word;
  ${props =>
    !props.noShadow &&
    css`
      box-shadow: 0px 10px 10px #e8e8e8;
    `}
`;

export class HoverablePreview extends React.Component<HoverablePreviewProps> {
  static Cell = HoverablePreviewCell;
  static Header = HoverablePreviewHeader;
  render() {
    return (
      <StyledHoverablePreview>{this.props.children}</StyledHoverablePreview>
    );
  }
}
