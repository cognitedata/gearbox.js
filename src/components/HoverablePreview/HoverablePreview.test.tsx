import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';
import { adjustBorders } from './components/HoverablePreviewCell/HoverablePreviewCell';
import { HoverablePreview } from './HoverablePreview';

configure({ adapter: new Adapter() });

const hpDisplayName = '.hp-displayicon';
const hpPreviewName = '.hp-preview';
const hpCellName = '.hp-cell';
const outsideName = '.outside';

describe('HoverablePreview basic', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <HoverablePreview>
          <HoverablePreview.Header>
            <HoverablePreview.Title>Test</HoverablePreview.Title>
            <HoverablePreview.CloseButton />
          </HoverablePreview.Header>
        </HoverablePreview>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('properly gives cell a title', () => {
    const wrapper = mount(
      <HoverablePreview>
        <HoverablePreview.Cell title="with title">
          Cell with title
        </HoverablePreview.Cell>
        <HoverablePreview.Cell>Cell without title</HoverablePreview.Cell>
      </HoverablePreview>
    );
    expect(
      wrapper
        .find(hpCellName)
        .at(0)
        .find('.hp-cell-title')
        .exists()
    ).toBeTruthy();
    expect(
      wrapper
        .find(hpCellName)
        .at(4)
        .find('.hp-cell-title')
        .exists()
    ).not.toBeTruthy();
  });
  it('animates correctly', () => {
    const wrapper = mount(
      <div className="test">
        Hover me
        <HoverablePreview displayOn="hover" fadeIn={true}>
          <HoverablePreview.Header>
            <HoverablePreview.Title>Test</HoverablePreview.Title>
            <HoverablePreview.CloseButton />
          </HoverablePreview.Header>
        </HoverablePreview>
      </div>
    );
    const displayIcon = wrapper.find(hpDisplayName).at(0);
    displayIcon.simulate('mouseover');

    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: true,
    });
  });
  it('removes shadow correctly', () => {
    const wrapper = mount(
      <div className="test">
        Hover me
        <HoverablePreview noShadow={true}>test</HoverablePreview>
      </div>
    );
    const element = wrapper.find(hpPreviewName).at(0);
    expect(
      getComputedStyle(element.getDOMNode()).getPropertyValue('box-shadow')
    ).toBe('');
  });
  it('properly aligns cells', () => {
    const wrapper = mount(
      <HoverablePreview>
        <HoverablePreview.Cell align="right">Right cell</HoverablePreview.Cell>
        <HoverablePreview.Cell>Center cell</HoverablePreview.Cell>
        <HoverablePreview.Cell align="left">Left cell</HoverablePreview.Cell>
      </HoverablePreview>
    );
    const propWidth = 'width';
    const propMargin = 'margin-left';

    const cellRight = wrapper.find(hpCellName).at(0);
    expect(
      getComputedStyle(cellRight.getDOMNode()).getPropertyValue(propWidth)
    ).toBe('50%');
    expect(
      getComputedStyle(cellRight.getDOMNode()).getPropertyValue(propMargin)
    ).toBe('auto');

    const cellCenter = wrapper.find(hpCellName).at(4);
    expect(
      getComputedStyle(cellCenter.getDOMNode()).getPropertyValue(propWidth)
    ).toBe('100%');
    expect(
      getComputedStyle(cellCenter.getDOMNode()).getPropertyValue(propMargin)
    ).toBe('');

    const cellLeft = wrapper.find(hpCellName).at(7);
    expect(
      getComputedStyle(cellLeft.getDOMNode()).getPropertyValue(propWidth)
    ).toBe('50%');
    expect(
      getComputedStyle(cellLeft.getDOMNode()).getPropertyValue(propMargin)
    ).toBe('');
  });
  it('adds required borders correctly', () => {
    const wrapper = mount(
      <div className="test">
        <HoverablePreview>
          <HoverablePreview.Cell borders={['top', 'bottom']}>
            Test
          </HoverablePreview.Cell>
        </HoverablePreview>
      </div>
    );

    const element = wrapper.find(hpCellName).at(0);
    expect(
      getComputedStyle(element.getDOMNode()).getPropertyValue('border-style')
    ).toBe('solid none solid none');
  });
  it('properly adjusts borders', () => {
    const onlyTop = adjustBorders(['top']);
    const leftRight = adjustBorders(['left', 'right']);
    const allFour = adjustBorders(['left', 'right', 'top', 'bottom']);

    expect(onlyTop).toContain('border-style: solid none none none;');
    expect(leftRight).toContain('border-style: none solid none solid;');
    expect(allFour).toContain('border-style: solid solid solid solid;');
  });
});

describe('HoverablePreview - events', () => {
  const events: any = {};

  beforeEach(() => {
    document.addEventListener = jest.fn((event, cb) => {
      events[event] = cb;
    });
  });
  afterEach(() => {
    document.removeEventListener = jest.fn((event, cb) => {
      events[event] = cb;
    });
  });

  it('appears on hover and perishes correctly', () => {
    const wrapper = mount(
      <div className="test">
        Hover me
        <HoverablePreview displayOn="hover">test</HoverablePreview>
      </div>
    );
    const instance: HoverablePreview = wrapper
      .find(HoverablePreview)
      .instance() as HoverablePreview;
    const isShown = jest.spyOn(instance, 'isShown');
    const displayIcon = wrapper.find(hpDisplayName).at(0);

    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: false,
    });

    displayIcon.simulate('mouseover');
    expect(isShown).toBeCalled();
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: true,
    });

    displayIcon.simulate('mouseleave');
    expect(isShown).toBeCalled();
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: false,
    });
  });

  it('appears on click and perishes correctly after clicking displayIcon again', () => {
    const wrapper = mount(
      <div>
        <div className="outside">Click me!</div>
        <HoverablePreview displayOn="click">test</HoverablePreview>
      </div>
    );
    const instance: HoverablePreview = wrapper
      .find(HoverablePreview)
      .instance() as HoverablePreview;
    const isShown = jest.spyOn(instance, 'isShown');
    const displayIcon = wrapper.find(hpDisplayName).at(0);
    const hpPreview = wrapper.find(hpPreviewName).at(0);

    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: false,
    });

    displayIcon.simulate('click');
    expect(isShown).toBeCalledTimes(1);
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: true,
    });

    hpPreview.simulate('click');
    expect(isShown).toBeCalledTimes(1);
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: true,
    });

    displayIcon.simulate('click');
    expect(isShown).toBeCalledTimes(2);
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: false,
    });
  });

  it('appears on click and perishes correctly after clicking outside', () => {
    const wrapper = mount(
      <div className="outside" style={{ width: '100vw', height: '100vh' }}>
        Click me!
        <HoverablePreview displayOn="click">test</HoverablePreview>
      </div>
    );
    const instance: HoverablePreview = wrapper
      .find(HoverablePreview)
      .instance() as HoverablePreview;
    const isShown = jest.spyOn(instance, 'isShown');
    const displayIcon = wrapper.find(hpDisplayName).at(0);

    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: false,
    });

    displayIcon.simulate('click');
    expect(isShown).toBeCalledTimes(1);
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: true,
    });

    events.click(outsideName);
    expect(isShown).toBeCalledTimes(2);
    expect(wrapper.find(HoverablePreview).state()).toMatchObject({
      show: false,
    });
  });
});
