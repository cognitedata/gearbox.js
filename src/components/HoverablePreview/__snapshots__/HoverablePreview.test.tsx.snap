// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverablePreview basic renders correctly 1`] = `
<div
  className="hp-wrapper sc-ifAKCX elMEqD"
>
  <div
    className="hp-displayicon sc-bxivhb kEsdhe"
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
  >
    <i
      aria-label="icon: question-circle"
      className="anticon anticon-question-circle"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="question-circle"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0 1 30.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1 0 80 0 40 40 0 1 0-80 0z"
        />
      </svg>
    </i>
  </div>
  <div
    className="hp-preview sc-EHOje fdVKtR"
  >
    <div
      className="hp-header sc-bwzfXH joaCVf"
    >
      <span
        className="hp-title sc-htpNat dDvXVh"
      >
        Test
      </span>
      <div
        className="hp-close sc-bxivhb kEsdhe"
      >
        <i
          aria-label="icon: close"
          className="anticon anticon-close"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </i>
      </div>
    </div>
  </div>
</div>
`;
